### PR TITLE
improvement(libs/header): Check for requested height before fetching it from the store

### PR DESCRIPTION
**Pull request by vgonkivs**
## Overview

Resolves https://github.com/celestiaorg/celestia-node/issues/1697

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: libs/header/store/store.go [hunk 1]
```diff
@@ -283,6 +283,10 @@ func (s *Store[H]) Has(ctx context.Context, hash header.Hash) (bool, error) {
 	return s.ds.Has(ctx, datastore.NewKey(hash.String()))
 }
 
+func (s *Store[H]) HasAt(height uint64) bool {
+	return height != uint64(0) && s.Height() >= height
+}
```
Comment (Wondertan) [inline libs/header/store/store.go:288]: Let's make ctx as a first param. 

Right now, we assume that the Store stores everything below `Height`. However, we will break this assumption once we will introduce pruning. Then we will do a disk look up which requires context and breaking the interface. Let's avoid breaking in it in future by introducing future proof form of it

**Review by Wondertan — COMMENTED** [review]
One request, and I think this is a very good idea to have

File: libs/header/store/store.go (hunk 1 above)
Comment (renaynay) [reply libs/header/store/store.go:288]: fine w me

File: libs/header/p2p/server.go [hunk 2]
```diff
@@ -201,7 +201,13 @@ func (serv *ExchangeServer[H]) handleRequest(from, to uint64) ([]H, error) {
 	}
 
 	log.Debugw("server: handling headers request", "from", from, "to", to)
-	headersByRange, err := serv.getter.GetRangeByHeight(ctx, from, to)
+	if !serv.store.HasAt(to - 1) {
+		span.SetStatus(codes.Error, header.ErrNotFound.Error())
+		log.Warnw("server: requested headers not stored")
```
Comment (renaynay) [inline libs/header/p2p/server.go:206]: I'd do a debug log as this can be more frequent - also log the `to` height

**Review by renaynay — APPROVED** [review]

File: libs/header/store/store.go (hunk 1 above)
Comment (vgonkivs) [reply libs/header/store/store.go:288]: true. Good idea, thank you.

File: libs/header/p2p/server.go (hunk 2 above)
Comment (vgonkivs) [reply libs/header/p2p/server.go:206]: Changed to debug. But I do not think we really need to print `to` height one more time. We are logging 
`log.Debugw("server: handling headers request", "from", from, "to", to)`  before `HasAt` check.

**Review by Wondertan — DISMISSED** [review]

File: libs/header/p2p/server.go (hunk 2 above)
Comment (renaynay) [reply libs/header/p2p/server.go]: those can always be lost though, meaning logs get cloudy so it's nice to know which request this is attributed to

**Review by renaynay — APPROVED** [review]

File: libs/header/p2p/server.go (hunk 2 above)
Comment (vgonkivs) [reply libs/header/p2p/server.go]: ok, np.

**Review by renaynay — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

**Pull request merged**